ndscan/experiment/utils.py

```diff
@@ -20,22 +22,26 @@ def is_kernel(func) -> bool:
     return meta.core_name is not None and not meta.portable
 
 
-class NumpyToVanillaEncoder(json.JSONEncoder):
+class Encoder(json.JSONEncoder):
     def default(self, obj):
         if isinstance(obj, numpy.integer):
             return int(obj)
         if isinstance(obj, numpy.floating):
             return float(obj)
         if isinstance(obj, numpy.ndarray):
             return obj.tolist()
+        if dataclasses.is_dataclass(obj):
```

Comment:
I don't dislike this (although the `dump_json` docstring would need adapting), but I wonder whether we can come up with some sort of best practices for using dataclasses/… in PYON. (I used JSON here instead of PYON on purpose to make it easier to load the result files from e.g. Mathematica or Julia, but that could be adapted.)

Reply:
> but I wonder whether we can come up with some sort of best practices for using dataclasses/… in PYON

You mean best practices within ndscan or best practices for general pyon? If we do go down the route of replacing the dictionary schema with something like dataclasses, I was wondering of having a `Schema` (or `SerialisableDataClass` or whatever) base class which helps to formalise some of this. That could then provide `to_dict` and `from_dict` (or `describe` and `decode` or whatever) methods to help keep things uniform.